### DISCUSSION TITLE
Changing between simulation models

### DISCUSSION THREAD

**Discussion by MarcusBinder** [NatLabRockies/floris#399] [Q&A] [answered]
Hello

I have a question regarding how to easy change between different simulation models. 

I am currently defining a model using a .Yaml file which includes wake model strings for the combination, deflection, turbulence model, and velocity deficit models.  
Is there an easy way to change these like I can with the fi.reinitialize(), to change things like e.g. air density or layout.
I would like to load in a model and run it with first one wake model, and then try to change to another and plot the difference. 

Also when looking into the source code I saw that line 26 in sosfs.py says :  model_string = "crespo_hernandez"
Is there a reason why this isn’t model_string = “sosfs”, or am I missing something? 

Thanks in advance!

**Comment by rafmudaf** [comment]
Hi @MarcusBinder 

The model string in the SOSFS combination model is incorrect, and I will fix it. However, in the case of the combination models it does not have an effect on the configuration of FLORIS because the inputs are mapped to the models through a mapping in the [Wake](https://github.com/NREL/floris/blob/main/floris/simulation/wake.py#L33) class.

As for changing the wake model in a script, I recommend to export the FLORIS settings as a dictionary, make the change you want, and then create a new FlorisInterface object passing this dictionary as the first argument.

```python
d_floris = fi.floris.as_dict()
d_floris["wake"]["model_strings"]["velocity_model"] = "cc"
d_floris["wake"]["wake_velocity_parameters"]["cc"] = {
    "a_s": 0.179367259,
    "b_s": 0.0118889215,
    "c_s1": 0.0563691592,
    "c_s2": 0.13290157,
    "a_f": 3.11,
    "b_f": -0.68,
    "c_f": 2.41,
    "alpha_mod": 1.0,
}
fi = FlorisInterface(d_floris)
```

This workflow is cumbersome, so we'll think about how to design this into the `floris.tools` package.

**Reply by MarcusBinder** [reply]
Thanks for the fast reply! 

That makes sense!

**Comment by Bartdoekemeijer** [comment, marked as answer]
Hi Marcus,

It seems that @rafmudaf already answered you! I was actually in the process of writing up a small demo script for you. I'll add it here nonetheless, in case you find it useful.

```
import copy
import matplotlib.pyplot as plt

from floris.tools import FlorisInterface
from floris.tools.visualization import visualize_cut_plane


if __name__ == "__main__":
    # Instantiate FLORIS using the GCH model and a 2-turbine wind farm
    fi = FlorisInterface("inputs/gch.yaml")
    fi.reinitialize(layout=[[0.0, 700.0], [0.0, 0.0]], wind_directions=[90.0])

    # Plot horizontal slice of GCH model
    fig, ax = plt.subplots(nrows=3, sharex=True, sharey=True)
    horizontal_plane_gch = fi.calculate_horizontal_plane(height=90.0)
    im = visualize_cut_plane(
        horizontal_plane_gch,
        minSpeed=0.0,
        maxSpeed=9.0,
        ax=ax[0],
        title="GCH model"
    )
    plt.colorbar(im, ax=ax[0])

    # Modify velocity deficit model to "CC" and update floris object (as by @rafmudaf's comment)
    fi_dict = fi.floris.as_dict()  # Export current floris configuration as dict
    fi_dict["wake"]["model_strings"]["velocity_model"] = "cc"  # Make changes
    fi = FlorisInterface(fi_dict)  # Initialize new floris class with updated dict

    # Plot horizontal slice of CC model
    horizontal_plane_cc = fi.calculate_horizontal_plane(height=90.0)
    im = visualize_cut_plane(
        horizontal_plane_cc,
        minSpeed=0.0,
        maxSpeed=9.0,
        ax=ax[1],
        title="CC model"
    )
    plt.colorbar(im, ax=ax[1])

    # Now create a CutPlane object with difference of the two models
    difference_hor_plane_gch_cc = copy.deepcopy(horizontal_plane_gch)
    difference_hor_plane_gch_cc.df[["u", "v", "w"]] = (
        horizontal_plane_gch.df[["u", "v", "w"]] -
        horizontal_plane_cc.df[["u", "v", "w"]]
    )
    im = visualize_cut_plane(
        difference_hor_plane_gch_cc,
        ax=ax[2],
        title="Difference (GCH minus CC)",
        minSpeed=-2.0,
        maxSpeed=2.0,
    )
    plt.colorbar(im, ax=ax[2])
    plt.show()
```

Which produces:
![image](https://user-images.githubusercontent.com/22119448/159521699-6f0c1d83-8f5a-41d1-bfec-5348cc7bf118.png)

**Reply by rafmudaf** [reply]
Oh the different plot is nice

**Reply by MarcusBinder** [reply]
Thanks for the sample code! That last plot is super cool!

**Comment by bayc** [comment]
One more addition to Bart's very helpful sample script, you can store the modified `fi` object in a new variable and keep it alongside the previous `fi` object if you would want to change and compare other things in the future. For example, you could store the new wake model `fi` object as `fi_cc`:

```
    # Modify velocity deficit model to "CC" and update floris object (as by @rafmudaf's comment)
    fi_dict = fi.floris.as_dict()  # Export current floris configuration as dict
    fi_dict["wake"]["model_strings"]["velocity_model"] = "cc"  # Make changes
    fi_cc = FlorisInterface(fi_dict)  # Initialize new floris class with updated dict
```

**Reply by Bartdoekemeijer** [reply]
Since @bayc 's post, a `copy` function has been added to `FlorisInterface`. Hence, you can create a copy of the current FlorisInterface object using
`fi_old = fi.copy()`
before then going on to create a modified object.

**Comment by wjscooty** [comment]
It is very useful!Thanks
